File: app/src/main/java/com/dmrandevu/gallery/media/blur/FaceFinder.kt
package com.dmrandevu.gallery.media.blur

import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions

/** Finds faces. */
class FaceFinder : RegionFinder {

    override val samplePeriodMs = BlurTimeline.SAMPLE_PERIOD_MS

    private val detector = FaceDetection.getClient(
        FaceDetectorOptions.Builder()
            // ACCURATE, not FAST: on real dashcam footage FAST walks past faces turned even
            // slightly away from the camera, and a missed face is the one failure this feature
            // cannot have.
            .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_ACCURATE)
            // Head width as a fraction of frame width; a bystander down the street still counts.
            .setMinFaceSize(0.05f)
            // Tracking deliberately left off. It is built for a live camera stream, and on a
            // decoded file it both suppresses detections and makes repeat runs over the same
            // video disagree with each other — measured at 140, 160 and 156 hits out of 238
            // samples across three runs of one clip. The timeline chains boxes by overlap
            // anyway, so the tracking ids bought little.
            .build()
    )

    override suspend fun regionsIn(frame: ScannedFrame): List<BlurTimeline.Region> {
        // Native size: a head is far bigger than the detector's minimum already, so there is
        // nothing to gain from enlarging the picture first.
        val image = frame.original()
        return detector.process(image).await().map {
            BlurTimeline.Region(it.boundingBox.normalizedIn(image), BlurTimeline.Shape.ELLIPSE)
        }
    }

    override fun close() = detector.close()
}

File: ios/DMRandevuGaleri/Media/Blur/FaceFinder.swift
import CoreGraphics
import Foundation
import Vision

/// Finds faces.
///
/// Vision's own detector replaces ML Kit here. It is the same class of thing — a bundled,
/// on-device face detector — and it needs no model file, so the plate model is the only weight
/// the app carries.
final class FaceFinder: RegionFinder {

    let samplePeriodMS = BlurTimeline.samplePeriodMS

    func regions(in frame: ScannedFrame) throws -> [BlurTimeline.Region] {
        let request = VNDetectFaceRectanglesRequest()
        // Revision 3 is the one that finds faces turned away from the camera. On real dashcam
        // footage the older revision walks straight past them, and a missed face is the one
        // failure this feature cannot have.
        request.revision = VNDetectFaceRectanglesRequestRevision3

        // Native size, and the buffer itself: a head is far bigger than the detector's floor
        // already, so there is nothing to gain from handing it a bigger picture. Vision is told
        // the orientation and answers in the upright space the timeline works in.
        let handler = VNImageRequestHandler(
            cvPixelBuffer: frame.pixelBuffer,
            orientation: frame.orientation,
            options: [:]
        )
        try handler.perform([request])

        return (request.results ?? []).map { face in
            BlurTimeline.Region(box: flippedToTopLeft(face.boundingBox), shape: .ellipse)
        }
    }
}
